frontend/lib/types/Type.cpp

```diff
@@ -337,6 +355,60 @@ bool Type::isPod(Context* context, const Type* t) {
   return true;
 }
 
+static bool const& isDefaultInitializableQuery(Context* context, const Type* t) {
+  QUERY_BEGIN(isDefaultInitializableQuery, context, t);
+
+  bool result = true;
+  if (!t || t->isUnknownType() || t->isErroneousType()) {
+    result = false;
+  } else if (t->isBuiltinType()) {
+    result = t->genericity() == Type::CONCRETE;
+  } else if (auto at = t->toArrayType()) {
+    result = isDefaultInitializableQuery(context, at->eltType().type());
+  } else if (t->isDomainType()) {
+    result = true; // production always returns true for domains.
+  } else if (t->isExternType()) {
+    // Currently extern records aren't initialized at all by default.
+    // But it's not necessarily reasonable to expect them to have
+    // initializers. See issue #7992 and preFold.cpp's setRecordDefaultValueFlags
+    // for FLAG_EXTERN.
+    result = true;
+  } else if (auto ct = t->toClassType()) {
+    result = ct->decorator().isNilable();
+  } else if (t->isTupleType()) {
+    result = checkFieldsWithPredicate(context, t, Type::isDefaultInitializable);
+  } else if (t->isRecordLike()) {
+    // If the type doesn't have a user-defined initializer or is a tuple, check
+    // its fields.
+    auto fieldsDefaultInitializable = true;
+    if (resolution::needCompilerGeneratedMethod(context, t, USTR("init"), /* parenless */ false)) {
+      fieldsDefaultInitializable = checkFieldsWithPredicate(context, t, Type::isDefaultInitializable);
+    }
+
+    if (!fieldsDefaultInitializable) {
+      result = false;
+    } else {
+      const uast::AstNode* ast = nullptr;
+      if (auto ct = t->getCompositeType()) {
+        if (auto id = ct->id()) {
+          ast = parsing::idToAst(context, std::move(id));
+        }
+      }
+
+      // note: production disallows default-init for generic fields like `var x;`,
+      // but why? Seems like this is an implementation detail. Allow it in Dyno.
```

Comment:
Doesn't that make sense? What value do you default init `var x` to?

Reply:
In other words, I don't see any philosophical reason why the following program shouldn't work:

```Chapel
record R {
  var x;
}

var x: R(int);
writeln(x);
```

Reply:
I've clarified the comment to say that default-init is disallowed even if the field is instantiated with a default-initable value, which gets to the root of my "but why?" question.